src/Initializer/ParameterDB.cpp

```diff
@@ -80,6 +84,83 @@ easi::Query seissol::initializers::ElementBarycentreGenerator::generate() const
   return query;
 }
 
+seissol::initializers::ElementAverageGenerator::ElementAverageGenerator(MeshReader const& meshReader)
+  : m_meshReader(meshReader)
+  {
+    // Generate subpoints and weights in reference tetrahedron using Gaussian quadrature
+    double quadraturePoints[NUM_QUADPOINTS][3];
+    double quadratureWeights[NUM_QUADPOINTS];
+    seissol::quadrature::TetrahedronQuadrature(quadraturePoints, quadratureWeights, CONVERGENCE_ORDER);
+    // Initialize const class members with results
+    std::copy(std::begin(quadratureWeights), std::end(quadratureWeights), std::begin(m_quadratureWeights));
+    for (int i = 0; i < NUM_QUADPOINTS; ++i) {
+      std::copy(std::begin(quadraturePoints[i]), std::end(quadraturePoints[i]), std::begin(m_quadraturePoints[i]));
+    }
+    // Initialize element volumes
+    m_elemVolumes = elementVolumes();
+  }
+
+easi::Query seissol::initializers::ElementAverageGenerator::generate() const {
+  std::vector<Element> const& elements = m_meshReader.getElements();
+  std::vector<Vertex> const& vertices = m_meshReader.getVertices();
+
+  // Generate query using subpoints for each element
+  easi::Query query(elements.size() * NUM_QUADPOINTS, 3);
+  
+  // Transform subpoints to global coordinates for all elements
+  for (unsigned elem = 0; elem < elements.size(); ++elem) {
```

Comment:
would a `#pragma omp parallel for` make sense?

Reply:
@sebwolf-de Out of my scope